examples/example_mcap_writer.cpp

```diff
@@ -0,0 +1,83 @@
+//
+// Copyright (c) 2024, Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
+// SPDX-License-Identifier: MPL-2.0
+//
+
+#include <osi-utilities/tracefile/writer/MCAPTraceFileWriter.h>
+
+#include <filesystem>
+
+#include "osi_sensorview.pb.h"
+#include "osi_version.pb.h"
+
+std::filesystem::path GenerateTempFilePath() {
+    return std::filesystem::temp_directory_path() / "sv_example.mcap";  // add sv to indicate sensor view as recommended by the OSI-specification
+}
+
+int main(int argc, const char** argv) {
+    std::cout << "Starting MCAP Writer example:" << std::endl;
+
+    // Create writer and open file
+    auto trace_file_writer = osi3::MCAPTraceFileWriter();
+    const auto trace_file_path = GenerateTempFilePath();
+    std::cout << "Creating trace_file at " << trace_file_path << std::endl;
+    trace_file_writer.Open(trace_file_path);
+
+    // add required and optional metadata to the net.asam.osi.trace metadata record
+    auto net_asam_osi_trace_metadata = osi3::MCAPTraceFileWriter::PrepareRequiredFileMetadata();
+    // Add optional metadata to the net.asam.osi.trace metadata record, as recommended by the OSI specification.
+    net_asam_osi_trace_metadata.metadata["description"] = "Example mcap trace file created with the ASAM OSI utilities library."; // optional field
+    net_asam_osi_trace_metadata.metadata["creation_time"] = osi3::MCAPTraceFileWriter::GetCurrentTimeAsString(); // optional field
+    net_asam_osi_trace_metadata.metadata["authors"] = "Jane Doe, John Doe"; // optional field
```

Comment:
Where is the rest of the mandatory metadata written/defined?

Reply:
Does the example fail? src/tracefile/writer/MCAPTraceFileWriter.cpp line 119 requires it?

```
        constexpr std::array<const char*, 5> kRequiredFields = {"version", "min_osi_version", "max_osi_version", "min_protobuf_version", "max_protobuf_version"};
        for (const auto& field : kRequiredFields) {
            if (metadata.metadata.find(field) == metadata.metadata.end()) {
                std::cerr << "ERROR: cannot add net.asam.osi.trac meta-data record without a " << field << " field.\n";
                return false;
            }
        }
```

Reply:
[Here](https://github.com/Lichtblick-Suite/asam-osi-utilities/pull/2/files/0df13a4432946a5a08f13f386e62d56c5f642e0f#diff-56a24a74e9814d13490bf27c09db2a08b133b14142bc35f935a798e37db98a5cR27) it gets created  and [here](https://github.com/Lichtblick-Suite/asam-osi-utilities/pull/2/files/0df13a4432946a5a08f13f386e62d56c5f642e0f#diff-56a24a74e9814d13490bf27c09db2a08b133b14142bc35f935a798e37db98a5cR32) the mandatory and optional metadata are written. 

The convenience function `osi3::MCAPTraceFileWriter::PrepareRequiredFileMetadata();` provides the required (by the current state of the new OSI spec) metadata. `PrepareRequiredFileMetadata()`  is obviously not useful for merging mcap files etc. as the min and max values of osi/protobuf version metadata will all get the same value (therefore a convenience function and not part of setting any metadata automatically).

Reply:
> Does the example fail? src/tracefile/writer/MCAPTraceFileWriter.cpp line 119 requires it?

No because it uses `osi3::MCAPTraceFileWriter::PrepareRequiredFileMetadata()`

```bash

$ git rev-parse HEAD
0df13a4432946a5a08f13f386e62d56c5f642e0f
$ git status
On branch 1-add-a-c++-trace-file-readerwriter
Your branch is up to date with 'origin/1-add-a-c++-trace-file-readerwriter'.

nothing to commit, working tree clean

$ ./cmake-build-debug/examples/example_mcap_writer 
Starting MCAP Writer example:
Creating trace_file at "/tmp/sv_example.mcap"
Finished MCAP Writer example
```

Reply:
ah ok, but then this makes no sense `metadata.metadata["version"] = osi_version_string; `because it is the version of the spec (we don't have yet but we would anticipate 3.8.0 which should include the multi tracefile spec.

Reply:
This should (after merge and release) be the osi version which is mentioned in the CMAKE

Reply:
> ah ok, but then this makes no sense metadata.metadata["version"] = osi_version_string; because it is the version of the spec (we don't have yet but we would anticipate 3.8.0 which should include the multi tracefile spec.

Ah you mean this should be hard-coded? Yes this makes sense. 

> ah ok, but then this makes no sense metadata.metadata["version"] = osi_version_string; because it is the version of the spec (we don't have yet but we would anticipate 3.8.0 which should include the multi tracefile spec.


In the cmake you can specify what ever osi you want but that is not related with the spec we are trying to fulfill here right? Updating the osi you build against does not automatically fix potential osi mcap schema issue which need to be fixed in the library itself.  

> we would anticipate 3.8.0

The assigned milestone is called 3.7.1

> This should (after merge and release)

Which merge and release? OSI muli trace file spec or this PR of asam-osi-utilities? I can imagine that asam-osi-utilities releases **could** be made quite complicated: multiple asam-osi-utilities versions for multiple osi versions. It might make sense to simply target the newest OSI version and always provide independent improvements and fixes - but that is a discussion for another day.  



**To sum it up**: If you do not disagree I would simply hardcode this to `3.7.0-a63f47d7b21bb44edafca176700d9cedf6b12d81` which represents the current values of Pieres last commit (the `VERSION` file in that branch + commit).